browser/brave_rewards/rewards_tab_helper.cc

```diff
@@ -22,85 +32,103 @@
 RewardsTabHelper::~RewardsTabHelper() {
-  if (rewards_service_)
+  if (rewards_service_) {
     rewards_service_->RemoveObserver(this);
+  }
 #if !BUILDFLAG(IS_ANDROID)
   BrowserList::RemoveObserver(this);
 #endif
 }
 
+void RewardsTabHelper::AddObserver(Observer* observer) {
+  observer_list_.AddObserver(observer);
+}
+
+void RewardsTabHelper::RemoveObserver(Observer* observer) {
+  observer_list_.RemoveObserver(observer);
+}
+
+void RewardsTabHelper::SetPublisherIdForTab(const std::string& publisher_id) {
+  if (publisher_id != publisher_id_) {
+    publisher_id_ = publisher_id;
+    for (auto& observer : observer_list_) {
+      observer.OnPublisherForTabUpdated(publisher_id_);
+    }
+  }
+}
+
 void RewardsTabHelper::DidFinishLoad(
     content::RenderFrameHost* render_frame_host,
     const GURL& validated_url) {
-  if (!rewards_service_ || render_frame_host->GetParent())
+  if (!rewards_service_ || render_frame_host->GetParent()) {
     return;
+  }
 
 #if BUILDFLAG(ENABLE_IPFS)
-  auto ipns_url = web_contents()->GetURL();
+  auto ipns_url = GetWebContents().GetLastCommittedURL();
   if (ipns_url.SchemeIs(ipfs::kIPNSScheme)) {
     rewards_service_->OnLoad(tab_id_, ipns_url);
     return;
   }
 #endif
+
   rewards_service_->OnLoad(tab_id_, validated_url);
 }
 
 void RewardsTabHelper::DidFinishNavigation(content::NavigationHandle* handle) {
-  if (!rewards_service_ || !handle->IsInMainFrame() ||
-      !handle->HasCommitted() || handle->IsDownload())
+  if (!handle->IsInMainFrame() || !handle->HasCommitted() ||
+      handle->IsDownload()) {
     return;
+  }
+
+  auto id = GetPublisherIdFromURL(GetWebContents().GetLastCommittedURL());
+  SetPublisherIdForTab(id ? *id : "");
+
+  MaybeSavePublisherInfo();
 
-  rewards_service_->OnUnload(tab_id_);
+  if (rewards_service_) {
+    rewards_service_->OnUnload(tab_id_);
+  }
 }
 
 void RewardsTabHelper::ResourceLoadComplete(
     content::RenderFrameHost* render_frame_host,
     const content::GlobalRequestID& request_id,
     const blink::mojom::ResourceLoadInfo& resource_load_info) {
-  if (!rewards_service_ || !render_frame_host)
+  if (!rewards_service_ || !render_frame_host) {
     return;
+  }
 
-  // TODO(nejczdovc): do we need to get anyother type then XHR??
   switch (resource_load_info.request_destination) {
-    // Formerly ResourceType::kMedia
     case network::mojom::RequestDestination::kAudio:
     case network::mojom::RequestDestination::kTrack:
     case network::mojom::RequestDestination::kVideo:
-    // Best match for ResourceType::kXhr (though, not limited to kXhr)
     case network::mojom::RequestDestination::kEmpty:
-    // Formerly ResourceType::kImage
     case network::mojom::RequestDestination::kImage:
     case network::mojom::RequestDestination::kScript:
       rewards_service_->OnXHRLoad(tab_id_, GURL(resource_load_info.final_url),
-                                  web_contents()->GetURL(),
+                                  GetWebContents().GetLastCommittedURL(),
```

Comment:
GetVisibleURL is equivalent to GetURL, not GetLastCommittedURL
GetLastCommittedURL may be what we want, but maybe we should make that change separately in case it changes the behavior in some cases?

Reply:
Good catch - updated these existing calls to use `GetVisibleURL`.